RFM9x.nut

```diff
@@ -0,0 +1,399 @@
+class RFM9x {
+    
+    _spi = null;
+    _cs = null;
+    _irqPin = null;
+    _receiveHandler = null;
+    _sendFinished = true;
+    _error = false;
+
+    static FIFO = 0x00;
+    
+    // SPI masks
+    static WRITE_MASK = 0x80;
+    static READ_MASK = 0x00;
+    
+    static REG_OP_MODE = 0x01;
+    // f_RF = ((X_OSC) * Fr_f ) / (2^19). Resolution is 61.035 Hz if X_OSC = 32MHz
+    static REG_FREQUENCY_HIGH = 0x06; // MSB (3 bytes long, contiguous)
+    static X_OSC = 0x1E84800; // 32 MHz
+    static FREQ_STEP = 61.035;
+    
+    // LoRa Mode
+    static CONFIG_BYTE = 0x80; 
+    
+    // CPHA and CPOL must be 0 
+    
+    static REG_PA_CONFIG = 0x09;
+    static REG_PA_RAMP = 0x0A;
+    static REG_OCP = 0x0B;
+    static REG_LNA = 0x0C;
+    static REG_FIFO_ADDR_PTR = 0x0D;
+    static REG_FIFO_TX_BASE_ADDR = 0x0E; // write base address in FIFO data buffer
+    static REG_FIFO_RX_BASE_ADDR = 0x0F; // read base address in FIFO data buffer
+    static REG_FIFO_RX_CURRENT_ADDR = 0x10; // Start address of last packet received
+    
+    static REG_IRQ_FLAGS_MASK = 0x11;
+    
+    // Interrupt Masks AND flags
+    static RX_TIMEOUT = 7;
+    static RX_DONE = 6;
+    static PAYLOAD_CRC_ERROR = 5;
+    static VALID_HEADER = 4;
+    static TX_DONE = 3;
+    static CAD_DONE = 2;
+    static FHSS_CHANGE_CHANNEL = 1;
+    static CAD_DETECTED = 0;
+    
+    // IRQ
+    static REG_IRQ_FLAGS = 0x12;
+    
+    static REG_RX_N_BYTES = 0x13; // number of bytes in most recent payload
+    
+    static REG_RX_HEADER_CNT_VALUE_MSB = 0x14;
+    static REG_RX_HEADER_CNT_VALUE_LSB = 0x15;
+    static REG_RX_PACKET_CNT_VALUE_MSB = 0x16;
+    static REG_RX_PACKET_CNT_VALUE_LSB = 0x17;
+    
+    static REG_MODEL_STAT = 0x18;
+    static REG_PKT_SNR_VALUE = 0x19;
+    static REG_PKT_RSSI_VALUE = 0x1A;
+    static REG_RSSI_VALUE = 0x1B;
+    static REG_HOP_CHANNEL = 0x1C;
+    static REG_MODEM_CONFIG1 = 0x1D;
+    static REG_MODEM_CONFIG2 = 0x1E;
+    static REG_MODEM_CONFIG3 = 0x26;
+    static REG_SYMB_TIMEOUT_LSB = 0x1F;
+    
+    static REG_PREAMBLE_MSB = 0x20;
+    static REG_PREAMBLE_LSB = 0x21;
+    static REG_PAYLOAD_LENGTH = 0x22;
+    static REG_MAX_PAYLOAD_LENGTH = 0x23;
+    static REG_HOP_PERIOD = 0x24;
+    static REG_FIFO_RX_BYTE_ADDR = 0x25;
+    
+    
+    // -------------------- Shared Registers ---------------------- //
+
+    static REG_DIO_MAPPING1 = 0x40;
+    static REG_DIO_MAPPING2 = 0x41;
+    static REG_VERSION = 0x42;
+    static REG_TCXO = 0x4B;
+    
+    // -------------------- Some Tables --------------------------- //
+    
+    static BWTABLE = {
```

Comment:
Please provide comments to each of the constants/tables/variables to make the code self contained.